### DISCUSSION TITLE
to upgrade HA LXC or not to upgrade, that’s the question

### DISCUSSION THREAD

**Discussion by LubomirH** [tteck/Proxmox#2341] [Show and tell]
It began at the end of 2022 when I started with HA and decided to save the precious resources on my Proxmox server and choose to go the path of LXC with HA core and use your script to install it.
But in summer 2023 I found out that I was stuck on the HA 2023.7 version due to the Python 3.10 version.
As I am no expert in the field of linux, proxmox, LXC, python and or it's virtualization techniques, I strugled quite few days to collect the steps which need to be done 
and eventually found an important piece of the puzzle to finally be able to achieve my goal to do an inplace upgrade of the Home Assistant Core LXC.

I am lucky to share the procedure I have used, but I am wondering if this could be a proper and future-proof way for anyone and what might be its benefits or disadvantages.

For me is the main benefit, that the LXC stays in place with the same ID, without extra effort to clone it.
I think the procedure could be part of ttecks update script or maybe this could be an extra upgrade script in his collection.

The HA LXC upgrade consists of 3 main steps:
**Prerequisites**
 not sure if this is crucial, but I upgraded the whole Proxmox server, 
 first to the latest Proxmox version (7.4-17) on Debian 11 (bullseye)
 and then did dist-upgrade the OS to Debian 12 (bookworm) and the last Proxmox version (8.1.3)
1. Upgrade OS of the HA LXC
	after this step the HA was left intact, still version 2023.7 on Python 3.10
2. Upgrade pyenv
	after this step the HA was also left intact, still version 2023.7 on Python 3.10
	the Python 3.11 was installed side by side to the 3.10 version
3. Upgrade venv + HA
	after this step the HA was running the version 2024.1.2 on Python 3.11
4. Clean up
	this step is pending, had no time to analyze the leftovers

thanks for any reply ;)

Lubomir

PS: thanks a lot for your Proxmox script library ;)

[LXC HomeAssistant - Upgrade 1.OS.cmd.txt](https://github.com/tteck/Proxmox/files/13932552/LXC.HomeAssistant.-.Upgrade.1.OS.cmd.txt)
[LXC HomeAssistant - Upgrade 2.PYENV.cmd.txt](https://github.com/tteck/Proxmox/files/13932553/LXC.HomeAssistant.-.Upgrade.2.PYENV.cmd.txt)
[LXC HomeAssistant - Upgrade 3.venv+HA.cmd.txt](https://github.com/tteck/Proxmox/files/13932554/LXC.HomeAssistant.-.Upgrade.3.venv%2BHA.cmd.txt)

**Comment by tteck** [comment]
1) Create a new Home Assistant Core LXC

2) Within the Proxmox shell execute
```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/misc/copy-data/home-assistant-core-copy-data-home-assistant-core.sh)"
```
(from https://github.com/tteck/Proxmox/blob/main/misc/copy-data/README.md)

3) Shutdown your original Home Assistant Core LXC, and transfer the MAC address to the new Home Assistant Core LXC, start new Home Assistant Core LXC.

Done

**Reply by tteck** [reply]
If you want the same CTID, after the above steps are complete, remove your old Home Assistant Core LXC, freeing up that CTID. 

Make a backup of the new Home Assistant Core LXC, then go to the storage where the backup is stored, select it and restore (do not select "Start after restore"). It will automatically assign the first unused CTID, which _should_ be your old Home Assistant Core LXC CTID. 

Shutdown and remove the Home Assistant Core LXC with the undesired CTID and then start the Home Assistant Core LXC with the desired CTID.

**Comment by tteck** [comment]
> ~Upgrade pyenv~

~Currently, `pyenv` is not needed with Debian 12   with Home Assistant 2024.02, `pyenv` is once again needed to obtain Python 3.12.1~

Starting fresh can help avoid potential issues that may arise from the accumulation of configurations, dependencies, or changes over time. It allows for a clean slate, potentially improving system performance and reducing the risk of conflicts or errors.
